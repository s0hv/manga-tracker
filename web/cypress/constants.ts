export const nextAuthSessionCookie = 'next-auth.session-token';
