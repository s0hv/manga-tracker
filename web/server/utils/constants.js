export const csrfMissing = 'Missing CSRF token';
export const isDev = process.env.NODE_ENV !== 'production';
export const isTest = process.env.NODE_ENV === 'test';
export const NO_GROUP = 1;
