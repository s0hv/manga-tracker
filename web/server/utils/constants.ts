export const csrfMissing = 'CSRF error. Modifying requests must come from the same origin.';
export const isDev = process.env.NODE_ENV !== 'production';
export const isTest = process.env.NODE_ENV === 'test';
export const NO_GROUP = 1;

export const isSecure = (process.env.NEXTAUTH_URL || '').startsWith('https://');
export const cookiePrefix = (process.env.NEXTAUTH_URL || '').startsWith('https://') ? '__Secure-' : '';
