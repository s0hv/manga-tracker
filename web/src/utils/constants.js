export const NotificationTypes = {
  DiscordWebhook: 1,
};

export const QueryKeys = {
  NotificationsList: 'notifications-list',
};
