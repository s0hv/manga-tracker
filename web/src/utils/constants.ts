export const NotificationTypes = {
  DiscordWebhook: 1,
  Webhook: 2,
} as const;

export const QueryKeys = {
  NotificationsList: ['notifications-list'],
  MangaServices: 'admin-manga-service',
  Services: ['services'],
} as const;
