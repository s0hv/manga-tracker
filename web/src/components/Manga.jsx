import {
  Button,
  Container,
  Grid,
  IconButton,
  Paper,
  Tab,
  Tabs,
  Tooltip,
  Typography,
} from '@material-ui/core';

import { makeStyles } from '@material-ui/core/styles';

import {
  Edit as EditIcon,
  Settings as SettingsIcon,
} from '@material-ui/icons';

import Link from 'next/link';
import React, { useCallback, useEffect, useMemo, useState } from 'react';
import { useCSRF } from '../utils/csrf';
import { useUser } from '../utils/useUser';
import { followUnfollow } from '../utils/utilities';
import ChapterList from './ChapterList';
import MangaAliases from './MangaAliases';
import MangaInfo from './MangaInfo';

import MangaSourceList from './MangaSourceList';
import ReleaseHeatmap from './ReleaseHeatmap';
import { TabPanelCustom } from './utils/TabPanelCustom';
import { getMangaReleases } from '../api/chapter';
import { MangaCover } from './MangaCover';

const verticalBreakpoint = 910;

const useStyles = makeStyles((theme) => ({
  title: {
    width: '85%',
    textAlign: 'left',
    paddingBottom: '10px',
    [theme.breakpoints.down('sm')]: {
      width: '100%',
    },
  },
  titleBar: {
    display: 'flex',
    justifyContent: 'space-between',
  },
  titleBarButtonsContainer: {
    display: 'flex',
    width: '15%',
    justifyContent: 'flex-end',
    [theme.breakpoints.down('xs')]: {
      flexFlow: 'column',
      width: '12%',
    },
  },
  details: {
    display: 'flex',
    flexFlow: 'row',
    [theme.breakpoints.down(verticalBreakpoint)]: {
      flexFlow: 'wrap',
      justifyContent: 'center',
    },
  },
  detailText: {
    marginLeft: '5px',
    [theme.breakpoints.down(verticalBreakpoint)]: {
      marginLeft: '3px',
    },
  },
  infoTable: {
    marginLeft: '30px',
    marginTop: '3px',
    [theme.breakpoints.down(verticalBreakpoint)]: {
      marginLeft: '20px',
    },
    [theme.breakpoints.down('xs')]: {
      marginLeft: '10px',
    },
  },
  sourceList: {
    marginLeft: '35px',
    [theme.breakpoints.down(verticalBreakpoint)]: {
      marginLeft: '23px',
    },
    [theme.breakpoints.down('xs')]: {
      marginLeft: '13px',
    },
  },
  followButton: {
    marginTop: theme.spacing(2),
    marginBottom: theme.spacing(2),
  },
  paper: {
    padding: '1em',
    minWidth: '440px',
  },
  infoGrid: {
    marginLeft: theme.spacing(4),
    width: 'fit-content',
    overflow: 'auto',
    [theme.breakpoints.down('xs')]: {
      marginLeft: '0px',
    },
  },
  rootGrid: {
    [theme.breakpoints.down('sm')]: {
      justifyContent: 'center',
    },
  },
}));


function Manga(props) {
  const {
    mangaData: {
      manga,
      services,
      aliases,
    },
    userFollows = [],
  } = props;

  const [releaseData, setReleaseData] = useState([]);
  const [activeTab, setActiveTab] = useState(0);
  const changeTab = useCallback((e, newVal) => setActiveTab(newVal), []);
  const csrf = useCSRF();

  useEffect(() => {
    getMangaReleases(manga.mangaId)
      .then(js => {
        setReleaseData(js);
      });
  }, [manga.mangaId]);

  const { isAuthenticated, user } = useUser();

  const classes = useStyles();
  const [editing, setEditing] = useState(false);
  const startEditing = useCallback(() => setEditing(!editing), [editing]);

  const serviceUrlFormats = useMemo(() => {
    const serviceMap = {};
    services.forEach(service => { serviceMap[service.serviceId] = service.urlFormat });
    return serviceMap;
  }, [services]);

  return (
    <Container maxWidth='lg' disableGutters>
      <Paper className={classes.paper}>
        <div className={classes.titleBar}>
          <Typography className={classes.title} variant='h4'>{manga.title}</Typography>
          {user?.admin && (
            <div className={classes.titleBarButtonsContainer}>
              <Link href={`/admin/manga/${manga.mangaId}`} passHref>
                <Tooltip title='Admin page'>
                  <IconButton>
                    <SettingsIcon />
                  </IconButton>
                </Tooltip>
              </Link>
              <Tooltip title='Edit chapters'>
                <IconButton onClick={startEditing}>
                  <EditIcon />
                </IconButton>
              </Tooltip>
            </div>
          )}
        </div>
        <div className={classes.details}>
          <a href={manga.mal} target='_blank' rel='noreferrer noopener' aria-label='myanimelist page of the manga'>
            <MangaCover
              url={manga.cover}
              alt={manga.title}
            />
          </a>
          <Grid
            container
            justifyContent='space-between'
            className={classes.rootGrid}
          >
            <Grid
              container
              direction='column'
              className={classes.infoGrid}
            >
              <MangaInfo mangaData={manga} />
              <MangaAliases aliases={aliases} />
            </Grid>
            <MangaSourceList
              classesProp={[classes.sourceList]}
              items={services}
              userFollows={userFollows}
              followUnfollow={(serviceId) => followUnfollow(csrf, manga.mangaId, serviceId)}
            />
          </Grid>
        </div>
        {isAuthenticated && (
          <Button
            variant='contained'
            color='primary'
            onClick={followUnfollow(csrf, manga.mangaId, null)}
            className={classes.followButton}
          >
            {userFollows.indexOf(null) < 0 ? 'Follow' : 'Unfollow'}
          </Button>
        )}

        <Tabs onChange={changeTab} value={activeTab}>
          <Tab label='Chapters' value={0} />
          <Tab label='Stats' value={1} />
        </Tabs>
        <TabPanelCustom value={activeTab} index={0} noRerenderOnChange>
          <ChapterList
            editable={editing}
            serviceUrlFormats={serviceUrlFormats}
            mangaId={manga.mangaId}
          />
        </TabPanelCustom>
        <TabPanelCustom value={activeTab} index={1} noRerenderOnChange>
          <ReleaseHeatmap
            dataRows={releaseData}
          />
        </TabPanelCustom>
      </Paper>
    </Container>
  );
}

export default Manga;
