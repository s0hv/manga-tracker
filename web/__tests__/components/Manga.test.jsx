import React from 'react';
import { render, screen, act } from '@testing-library/react';
import fetchMock from 'fetch-mock';

import {
  adminUser,
  mockUTCDates,
  normalUser,
  withUser,
  mockNotistackHooks,
} from '../utils';
import Manga from '../../src/components/Manga';
import { emptyFullManga as emptyManga, fullManga as manga } from '../constants';

describe('Manga page should render correctly', () => {
  mockUTCDates();
  mockNotistackHooks();
  fetchMock.get('express:/api/manga/:mangaId/chapters', []);
  fetchMock.get('express:/api/chapter/releases/:mangaId', []);

  const follows = [1];

  /**
   * Based on a condition checks whether an element should or
   * should not exist in the document
   * @param {boolean} shouldExist Condition, which if true, makes sure that the element exists
   * @param {import('@testing-library/react').ByRoleMatcher} role The role of the element
   * @param {import('@testing-library/react').ByRoleOptions} options
   */
  const conditionalInDocument = (shouldExist, role, options) => {
    if (shouldExist) {
      expect(screen.getByRole(role, options)).toBeInTheDocument();
    } else {
      expect(screen.queryByRole(role, options)).not.toBeInTheDocument();
    }
  };

  const expectTitleExists = (m) => {
    expect(
      screen.getByRole('heading', { name: m.manga.title })
    ).toBeInTheDocument();
  };

  const expectMangaInfoExists = (m) => {
    // With cover mal link and the cover should exist. Otherwise they should be hidden
    if (m.manga.cover) {
      const malLink = screen.getByRole('link', { name: /myanimelist page of the manga/i });
      expect(malLink).toBeInTheDocument();
      expect(malLink).toHaveAttribute('href', m.manga.mal);

      const coverImage = screen.getByRole('img', { name: m.manga.title });
      expect(coverImage).toBeInTheDocument();
      expect(coverImage).toHaveAttribute('src', `${m.manga.cover}.256.jpg`);
    } else {
      expect(screen.queryByRole('link', { name: /myanimelist page of the manga/i })).not.toBeInTheDocument();
      expect(screen.queryByRole('img', { name: m.manga.title })).not.toBeInTheDocument();
    }

    if (m.aliases?.length > 0) {
      // Make sure the aliases are listed
      expect(screen.getByText(/alternative titles/i)).toBeInTheDocument();
      expect(screen.getByText(m.aliases[0], { exact: true })).toBeInTheDocument();
    } else {
      expect(screen.queryByText(/alternative titles/i)).not.toBeInTheDocument();
    }

    expect(screen.getByRole('table', { name: /manga information/i })).toBeInTheDocument();
  };

  const expectSourcesExist = (_) => {
    expect(screen.getByRole('list', { name: 'manga sources' })).toBeInTheDocument();
  };

  const expectAdminControls = (isAdmin) => {
    conditionalInDocument(isAdmin, 'link', { name: /admin page/i });

    conditionalInDocument(isAdmin, 'button', { name: /edit chapters/i });
  };

  const expectAuthenticatedFeatures = (isAuthenticated) => {
    conditionalInDocument(isAuthenticated, 'button', { name: /follow all releases/i });
  };

  it('should render correctly', async () => {
    await act(async () => {
      render(<Manga mangaData={manga} follows={follows} />);
    });

    expectTitleExists(manga);
    expectMangaInfoExists(manga);
    expectSourcesExist(manga);
    expectAdminControls(false);
    expectAuthenticatedFeatures(false);
  });

  it('should render correctly when logged in', async () => {
    await act(async () => {
      render(await withUser(
        normalUser,
        <Manga mangaData={manga} follows={follows} />
      ));
    });

    expectTitleExists(manga);
    expectMangaInfoExists(manga);
    expectSourcesExist(manga);
    expectAdminControls(false);
    expectAuthenticatedFeatures(true);
  });

  it('Should render correctly as admin', async () => {
    await act(async () => {
      render(await withUser(
        adminUser,
        <Manga mangaData={manga} follows={follows} />
      ));
    });

    expectTitleExists(manga);
    expectMangaInfoExists(manga);
    expectSourcesExist(manga);
    expectAdminControls(true);
    expectAuthenticatedFeatures(true);
  });

  it('should render correctly with little data', async () => {
    await act(async () => {
      render(<Manga mangaData={emptyManga} follows={follows} />);
    });

    expectTitleExists(emptyManga);
    expectMangaInfoExists(emptyManga);
    expectSourcesExist(emptyManga);
    expectAdminControls(false);
    expectAuthenticatedFeatures(false);
  });
});
