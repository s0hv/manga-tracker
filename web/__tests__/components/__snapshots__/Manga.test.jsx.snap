// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Manga page should render correctly Should render correctly as admin 1`] = `
<WithStyles(ForwardRef(Container))
  disableGutters={true}
  maxWidth="lg"
>
  <WithStyles(ForwardRef(Paper))
    className="makeStyles-paper-9"
  >
    <div
      className="makeStyles-titleBar-2"
    >
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-title-1"
        variant="h4"
      >
        Dr. STONE
      </WithStyles(ForwardRef(Typography))>
    </div>
    <div
      className="makeStyles-details-4"
    >
      <a
        aria-label="myanimelist page of the manga"
        href="https://myanimelist.net/manga/103897"
        rel="noreferrer noopener"
        target="_blank"
      >
        <MangaCover
          alt="Dr. STONE"
          url="/images/test.png"
        />
      </a>
      <WithStyles(ForwardRef(Grid))
        className="makeStyles-rootGrid-11"
        container={true}
        justifyContent="space-between"
      >
        <WithStyles(ForwardRef(Grid))
          className="makeStyles-infoGrid-10"
          container={true}
          direction="column"
        >
          <MangaInfo
            mangaData={
              Object {
                "al": "https://anilist.co/manga/98416",
                "amz": "https://www.amazon.co.jp/gp/product/B075F8JBQ1",
                "ap": "https://www.anime-planet.com/manga/dr-stone",
                "artist": "Boichi",
                "author": "Inagaki Riichiro",
                "bw": "https://bookwalker.jp/series/114645",
                "cover": "/images/test.png",
                "ebj": "https://www.ebookjapan.jp/ebj/413780/",
                "engtl": "https://www.viz.com/dr-stone",
                "estimatedRelease": "2020-07-12T16:00:00.000Z",
                "kt": "https://kitsu.io/manga/38860",
                "lastUpdated": "2020-06-28T08:15:55.170Z",
                "latestChapter": 157,
                "latestRelease": "2020-07-05T16:00:00.000Z",
                "mal": "https://myanimelist.net/manga/103897",
                "mangaId": 1,
                "mu": "https://www.mangaupdates.com/series.html?id=139601",
                "nu": "https://www.novelupdates.com/series/null",
                "raw": "null",
                "releaseInterval": Object {
                  "days": 7,
                },
                "status": 0,
                "title": "Dr. STONE",
              }
            }
          />
          <MangaAliases
            aliases={
              Array [
                "test alias",
              ]
            }
          />
        </WithStyles(ForwardRef(Grid))>
        <MangaSourceList
          classesProp={
            Array [
              "makeStyles-sourceList-7",
            ]
          }
          followUnfollow={[Function]}
          items={
            Array [
              Object {
                "name": "MANGA Plus",
                "serviceId": 1,
                "titleId": "100010",
                "url": "https://mangaplus.shueisha.co.jp/titles/{}",
                "urlFormat": "https://mangaplus.shueisha.co.jp/viewer/{}",
              },
              Object {
                "name": "MangaDex",
                "serviceId": 2,
                "titleId": "20882",
                "url": "https://mangadex.org/title/{}",
                "urlFormat": "https://mangadex.org/chapter/{}",
              },
            ]
          }
          userFollows={Array []}
        />
      </WithStyles(ForwardRef(Grid))>
    </div>
    <WithStyles(ForwardRef(Tabs))
      onChange={[Function]}
      value={0}
    >
      <WithStyles(ForwardRef(Tab))
        label="Chapters"
        value={0}
      />
      <WithStyles(ForwardRef(Tab))
        label="Stats"
        value={1}
      />
    </WithStyles(ForwardRef(Tabs))>
    <TabPanelCustom
      index={0}
      noRerenderOnChange={true}
      value={0}
    >
      <ChapterList
        editable={false}
        mangaId={1}
        serviceUrlFormats={
          Object {
            "1": "https://mangaplus.shueisha.co.jp/viewer/{}",
            "2": "https://mangadex.org/chapter/{}",
          }
        }
      />
    </TabPanelCustom>
    <TabPanelCustom
      index={1}
      noRerenderOnChange={true}
      value={0}
    >
      <ReleaseHeatmap
        dataRows={Array []}
      />
    </TabPanelCustom>
  </WithStyles(ForwardRef(Paper))>
</WithStyles(ForwardRef(Container))>
`;

exports[`Manga page should render correctly should render correctly 1`] = `
<WithStyles(ForwardRef(Container))
  disableGutters={true}
  maxWidth="lg"
>
  <WithStyles(ForwardRef(Paper))
    className="makeStyles-paper-9"
  >
    <div
      className="makeStyles-titleBar-2"
    >
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-title-1"
        variant="h4"
      >
        Dr. STONE
      </WithStyles(ForwardRef(Typography))>
    </div>
    <div
      className="makeStyles-details-4"
    >
      <a
        aria-label="myanimelist page of the manga"
        href="https://myanimelist.net/manga/103897"
        rel="noreferrer noopener"
        target="_blank"
      >
        <MangaCover
          alt="Dr. STONE"
          url="/images/test.png"
        />
      </a>
      <WithStyles(ForwardRef(Grid))
        className="makeStyles-rootGrid-11"
        container={true}
        justifyContent="space-between"
      >
        <WithStyles(ForwardRef(Grid))
          className="makeStyles-infoGrid-10"
          container={true}
          direction="column"
        >
          <MangaInfo
            mangaData={
              Object {
                "al": "https://anilist.co/manga/98416",
                "amz": "https://www.amazon.co.jp/gp/product/B075F8JBQ1",
                "ap": "https://www.anime-planet.com/manga/dr-stone",
                "artist": "Boichi",
                "author": "Inagaki Riichiro",
                "bw": "https://bookwalker.jp/series/114645",
                "cover": "/images/test.png",
                "ebj": "https://www.ebookjapan.jp/ebj/413780/",
                "engtl": "https://www.viz.com/dr-stone",
                "estimatedRelease": "2020-07-12T16:00:00.000Z",
                "kt": "https://kitsu.io/manga/38860",
                "lastUpdated": "2020-06-28T08:15:55.170Z",
                "latestChapter": 157,
                "latestRelease": "2020-07-05T16:00:00.000Z",
                "mal": "https://myanimelist.net/manga/103897",
                "mangaId": 1,
                "mu": "https://www.mangaupdates.com/series.html?id=139601",
                "nu": "https://www.novelupdates.com/series/null",
                "raw": "null",
                "releaseInterval": Object {
                  "days": 7,
                },
                "status": 0,
                "title": "Dr. STONE",
              }
            }
          />
          <MangaAliases
            aliases={
              Array [
                "test alias",
              ]
            }
          />
        </WithStyles(ForwardRef(Grid))>
        <MangaSourceList
          classesProp={
            Array [
              "makeStyles-sourceList-7",
            ]
          }
          followUnfollow={[Function]}
          items={
            Array [
              Object {
                "name": "MANGA Plus",
                "serviceId": 1,
                "titleId": "100010",
                "url": "https://mangaplus.shueisha.co.jp/titles/{}",
                "urlFormat": "https://mangaplus.shueisha.co.jp/viewer/{}",
              },
              Object {
                "name": "MangaDex",
                "serviceId": 2,
                "titleId": "20882",
                "url": "https://mangadex.org/title/{}",
                "urlFormat": "https://mangadex.org/chapter/{}",
              },
            ]
          }
          userFollows={Array []}
        />
      </WithStyles(ForwardRef(Grid))>
    </div>
    <WithStyles(ForwardRef(Tabs))
      onChange={[Function]}
      value={0}
    >
      <WithStyles(ForwardRef(Tab))
        label="Chapters"
        value={0}
      />
      <WithStyles(ForwardRef(Tab))
        label="Stats"
        value={1}
      />
    </WithStyles(ForwardRef(Tabs))>
    <TabPanelCustom
      index={0}
      noRerenderOnChange={true}
      value={0}
    >
      <ChapterList
        editable={false}
        mangaId={1}
        serviceUrlFormats={
          Object {
            "1": "https://mangaplus.shueisha.co.jp/viewer/{}",
            "2": "https://mangadex.org/chapter/{}",
          }
        }
      />
    </TabPanelCustom>
    <TabPanelCustom
      index={1}
      noRerenderOnChange={true}
      value={0}
    >
      <ReleaseHeatmap
        dataRows={Array []}
      />
    </TabPanelCustom>
  </WithStyles(ForwardRef(Paper))>
</WithStyles(ForwardRef(Container))>
`;

exports[`Manga page should render correctly should render correctly when logged in 1`] = `
<WithStyles(ForwardRef(Container))
  disableGutters={true}
  maxWidth="lg"
>
  <WithStyles(ForwardRef(Paper))
    className="makeStyles-paper-9"
  >
    <div
      className="makeStyles-titleBar-2"
    >
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-title-1"
        variant="h4"
      >
        Dr. STONE
      </WithStyles(ForwardRef(Typography))>
    </div>
    <div
      className="makeStyles-details-4"
    >
      <a
        aria-label="myanimelist page of the manga"
        href="https://myanimelist.net/manga/103897"
        rel="noreferrer noopener"
        target="_blank"
      >
        <MangaCover
          alt="Dr. STONE"
          url="/images/test.png"
        />
      </a>
      <WithStyles(ForwardRef(Grid))
        className="makeStyles-rootGrid-11"
        container={true}
        justifyContent="space-between"
      >
        <WithStyles(ForwardRef(Grid))
          className="makeStyles-infoGrid-10"
          container={true}
          direction="column"
        >
          <MangaInfo
            mangaData={
              Object {
                "al": "https://anilist.co/manga/98416",
                "amz": "https://www.amazon.co.jp/gp/product/B075F8JBQ1",
                "ap": "https://www.anime-planet.com/manga/dr-stone",
                "artist": "Boichi",
                "author": "Inagaki Riichiro",
                "bw": "https://bookwalker.jp/series/114645",
                "cover": "/images/test.png",
                "ebj": "https://www.ebookjapan.jp/ebj/413780/",
                "engtl": "https://www.viz.com/dr-stone",
                "estimatedRelease": "2020-07-12T16:00:00.000Z",
                "kt": "https://kitsu.io/manga/38860",
                "lastUpdated": "2020-06-28T08:15:55.170Z",
                "latestChapter": 157,
                "latestRelease": "2020-07-05T16:00:00.000Z",
                "mal": "https://myanimelist.net/manga/103897",
                "mangaId": 1,
                "mu": "https://www.mangaupdates.com/series.html?id=139601",
                "nu": "https://www.novelupdates.com/series/null",
                "raw": "null",
                "releaseInterval": Object {
                  "days": 7,
                },
                "status": 0,
                "title": "Dr. STONE",
              }
            }
          />
          <MangaAliases
            aliases={
              Array [
                "test alias",
              ]
            }
          />
        </WithStyles(ForwardRef(Grid))>
        <MangaSourceList
          classesProp={
            Array [
              "makeStyles-sourceList-7",
            ]
          }
          followUnfollow={[Function]}
          items={
            Array [
              Object {
                "name": "MANGA Plus",
                "serviceId": 1,
                "titleId": "100010",
                "url": "https://mangaplus.shueisha.co.jp/titles/{}",
                "urlFormat": "https://mangaplus.shueisha.co.jp/viewer/{}",
              },
              Object {
                "name": "MangaDex",
                "serviceId": 2,
                "titleId": "20882",
                "url": "https://mangadex.org/title/{}",
                "urlFormat": "https://mangadex.org/chapter/{}",
              },
            ]
          }
          userFollows={Array []}
        />
      </WithStyles(ForwardRef(Grid))>
    </div>
    <WithStyles(ForwardRef(Tabs))
      onChange={[Function]}
      value={0}
    >
      <WithStyles(ForwardRef(Tab))
        label="Chapters"
        value={0}
      />
      <WithStyles(ForwardRef(Tab))
        label="Stats"
        value={1}
      />
    </WithStyles(ForwardRef(Tabs))>
    <TabPanelCustom
      index={0}
      noRerenderOnChange={true}
      value={0}
    >
      <ChapterList
        editable={false}
        mangaId={1}
        serviceUrlFormats={
          Object {
            "1": "https://mangaplus.shueisha.co.jp/viewer/{}",
            "2": "https://mangadex.org/chapter/{}",
          }
        }
      />
    </TabPanelCustom>
    <TabPanelCustom
      index={1}
      noRerenderOnChange={true}
      value={0}
    >
      <ReleaseHeatmap
        dataRows={Array []}
      />
    </TabPanelCustom>
  </WithStyles(ForwardRef(Paper))>
</WithStyles(ForwardRef(Container))>
`;

exports[`Manga page should render correctly should render correctly with little data 1`] = `
<WithStyles(ForwardRef(Container))
  disableGutters={true}
  maxWidth="lg"
>
  <WithStyles(ForwardRef(Paper))
    className="makeStyles-paper-9"
  >
    <div
      className="makeStyles-titleBar-2"
    >
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-title-1"
        variant="h4"
      >
        Dr. STONE 2
      </WithStyles(ForwardRef(Typography))>
    </div>
    <div
      className="makeStyles-details-4"
    >
      <a
        aria-label="myanimelist page of the manga"
        href={null}
        rel="noreferrer noopener"
        target="_blank"
      >
        <MangaCover
          alt="Dr. STONE 2"
          url={null}
        />
      </a>
      <WithStyles(ForwardRef(Grid))
        className="makeStyles-rootGrid-11"
        container={true}
        justifyContent="space-between"
      >
        <WithStyles(ForwardRef(Grid))
          className="makeStyles-infoGrid-10"
          container={true}
          direction="column"
        >
          <MangaInfo
            mangaData={
              Object {
                "al": "https://anilist.co/manga/98416",
                "amz": "https://www.amazon.co.jp/gp/product/B075F8JBQ1",
                "ap": "https://www.anime-planet.com/manga/dr-stone",
                "artist": "Boichi",
                "author": "Inagaki Riichiro",
                "bw": "https://bookwalker.jp/series/114645",
                "cover": null,
                "ebj": "https://www.ebookjapan.jp/ebj/413780/",
                "engtl": "https://www.viz.com/dr-stone",
                "estimatedRelease": null,
                "kt": "https://kitsu.io/manga/38860",
                "lastUpdated": null,
                "latestChapter": null,
                "latestRelease": null,
                "mal": null,
                "mangaId": 2,
                "mu": "https://www.mangaupdates.com/series.html?id=139601",
                "nu": "https://www.novelupdates.com/series/null",
                "raw": "null",
                "releaseInterval": undefined,
                "status": 0,
                "title": "Dr. STONE 2",
              }
            }
          />
          <MangaAliases />
        </WithStyles(ForwardRef(Grid))>
        <MangaSourceList
          classesProp={
            Array [
              "makeStyles-sourceList-7",
            ]
          }
          followUnfollow={[Function]}
          items={
            Array [
              Object {
                "name": "MANGA Plus",
                "serviceId": 1,
                "titleId": "100010",
                "url": "https://mangaplus.shueisha.co.jp/titles/{}",
                "urlFormat": "https://mangaplus.shueisha.co.jp/viewer/{}",
              },
            ]
          }
          userFollows={Array []}
        />
      </WithStyles(ForwardRef(Grid))>
    </div>
    <WithStyles(ForwardRef(Tabs))
      onChange={[Function]}
      value={0}
    >
      <WithStyles(ForwardRef(Tab))
        label="Chapters"
        value={0}
      />
      <WithStyles(ForwardRef(Tab))
        label="Stats"
        value={1}
      />
    </WithStyles(ForwardRef(Tabs))>
    <TabPanelCustom
      index={0}
      noRerenderOnChange={true}
      value={0}
    >
      <ChapterList
        editable={false}
        mangaId={2}
        serviceUrlFormats={
          Object {
            "1": "https://mangaplus.shueisha.co.jp/viewer/{}",
          }
        }
      />
    </TabPanelCustom>
    <TabPanelCustom
      index={1}
      noRerenderOnChange={true}
      value={0}
    >
      <ReleaseHeatmap
        dataRows={Array []}
      />
    </TabPanelCustom>
  </WithStyles(ForwardRef(Paper))>
</WithStyles(ForwardRef(Container))>
`;
